ci/github-script/labels: auto close package request issues

wolfgangwalther opened the pull request:
This allows the labels workflow to support issue management in two ways:
- New package request can potentially created with a `4.workflow: auto-close` label immediately and be closed automatically this way. Plays together with #434609.
- Existing package requests can be bulk-closed by adding this label. This has the advantage of posting the explanatory comment at the same time, which is not possible with regular bulk operations.

Context of this PR is #425040.

## Things done

- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
ci/github-script/labels: auto close package request issues

This allows the labels workflow to support issue management in two ways:
- New package request can potentially created with a `4.workflow:
auto-close` label immediately and be closed automatically this way.
- Existing package requests can be bulk-closed by adding this label.
This has the advantage of posting the explanatory comment at the same
time, which is not possible with regular bulk operations.

SigmaSquadron commented:
When this label is added to all package requests, can we skip the package requests labelled with "5.scope: tracking"? They often have important discussions that go beyond just requests for random leaf packages.

SigmaSquadron reviewed (approved):
Seems like a plan!

MattSturgeon commented on ci/github-script/labels.js:
```diff
@@ -172,6 +172,40 @@ module.exports = async function ({ github, context, core, dry }) {
+
+    if (item.labels.some(({ name }) => name === '0.kind: packaging request')) {
+      const body = [
+        'Thanks for opening this package request. Unfortunately, in order to mitigate the unsustainable growth of unmaintained packages, Nixpkgs is no longer using package requests to coordinate new package additions.',
+        '',
+        'That said, **we encourage you to [contribute](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#quick-start-to-adding-a-package) this package as a Pull Request yourself** - should you need any help during packaging, feel free to reach out on [Matrix](https://matrix.to/#/#dev:nixos.org) or [Discourse](https://discourse.nixos.org/c/dev/14). You are also welcome to open custom _Tr ...
+        '',
+        '*This change has the benefit of decreasing future maintainer burden on the long term, and was implemented after several contributors reached a consensus. You can read more about this on Pull Request [391926](https://www.github.com/NixOS/nixpkgs/issues/391926) and Issue [425040](https://www.github.com/NixOS/nixpkgs/issues/425040).*',
```
I don't see any dynamic interpolation (etc). Would it be more maintainable to store this text in a dedicated file?

We already need a `ci` checkout to have access to this script, anyway.

MattSturgeon replied:
Also: where possible I prefer to use reference-style links. As per https://github.com/NixOS/nixpkgs/pull/434609#discussion_r2288910748

Also: some of this phrasing is the same as #434609. If my nit-picks are accepted there, then the wording here should be kept in sync 😅

MattSturgeon commented on ci/github-script/labels.js:
```diff
@@ -191,6 +225,9 @@ module.exports = async function ({ github, context, core, dry }) {
       } else {
         stats.issues++
+        if (item.labels.some(({ name }) => name === '4.workflow: auto-close')) {
+          if (await handleAutoClose(item)) return
```
Q: what's the early return for here?

Maybe `handleAutoClose` should document when/why it returns true/false?

MattSturgeon reviewed (approved):
Diff SGTM

wolfgangwalther committed:
ISSUE_TEMPLATE: revert one-sentence-per-line for package_request

We tried, but GitHub's markdown renderer will render these newlines
there, so going back to "real" paragraphs.

Also fixed the emphasis slightly, since "via Issues" is not bold, "via a
Pull Request" is also made not bold for consistency. Puts the focus more
on the "contribute" link as well.
